pkg/scheduler/preemption/fairsharing/ordering.go

```diff
@@ -0,0 +1,195 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairsharing
+
+import (
+	"iter"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// TargetClusterQueueOrdering defines an ordering over ClusterQueues
+// to consider during preemption.  This is done by traversing from
+// root Cohort, selecting the child which satisfies the 3 conditions:
+// 1. has candidate workloads
+// 2. has DRS >= 0
+// 3. has the highest DominantResourceShare of nodes passing filters 1 and 2.
+//
+// The same TargetClusterQueue may be returned multiple times in a
+// row, if its AlmostLeastCommonAncestor's DominantResourceShare is
+// still the highest.
+//
+// To guarantee progression, TargetClusterQueue.PopWorkload, or
+// TargetClusterQueueOrdering.DropQueue must be called between each
+// entry returned.
+type TargetClusterQueueOrdering struct {
+	preemptorCq *cache.ClusterQueueSnapshot
+	// ancestor Cohorts of the preemptor ClusterQueue.
+	preemptorAncestors sets.Set[*cache.CohortSnapshot]
+
+	clusterQueueToTarget map[kueue.ClusterQueueReference][]*workload.Info
+
+	// pruned nodes are nodes which we are certain will never
+	// yield more preemption target candidates. We use this set to
+	// determine our stopping condition: once the rootCohort is in
+	// the prunedCohorts list, we will not find any more
+	// preemption target candidates.
+	prunedClusterQueues sets.Set[*cache.ClusterQueueSnapshot]
+	prunedCohorts       sets.Set[*cache.CohortSnapshot]
+}
+
+func MakeClusterQueueOrdering(cq *cache.ClusterQueueSnapshot, candidates []*workload.Info) TargetClusterQueueOrdering {
+	t := TargetClusterQueueOrdering{
+		preemptorCq:        cq,
+		preemptorAncestors: sets.New[*cache.CohortSnapshot](),
+
+		clusterQueueToTarget: make(map[kueue.ClusterQueueReference][]*workload.Info),
+
+		prunedClusterQueues: sets.New[*cache.ClusterQueueSnapshot](),
+		prunedCohorts:       sets.New[*cache.CohortSnapshot](),
+	}
+
+	cohort := cq.Parent()
+	for cohort != nil {
+		t.preemptorAncestors.Insert(cohort)
+		cohort = cohort.Parent()
+	}
+
+	for _, candidate := range candidates {
+		t.clusterQueueToTarget[candidate.ClusterQueue] = append(t.clusterQueueToTarget[candidate.ClusterQueue], candidate)
+	}
+
+	return t
+}
+
+func (t *TargetClusterQueueOrdering) Iter() iter.Seq[*TargetClusterQueue] {
+	return func(yield func(v *TargetClusterQueue) bool) {
+		// handle CQ without Cohort case.
+		if !t.preemptorCq.HasParent() {
+			targetCq := &TargetClusterQueue{
+				ordering: t,
+				targetCq: t.preemptorCq,
+			}
+
+			for targetCq.HasWorkload() {
+				if !yield(targetCq) {
+					return
+				}
```

Comment:
Can't this lead to infinite loops?

Reply:
As long as the user follows the API contract - see TargetClusterQueueOrdering doc - it won't.

In general, this code worries me as an infinite loop risk. What do you think of having a max iterations breakout, so at least we'll print a debugging error message, so users can file an issue if there are any branches we miss? Rather than going into a livelock

Reply:
Yeah, it makes sense to add some safety mechanism maybe, cap it by the number of workloads?